File: go/03_函数/01_函数基础.go
// ============================================================
// Go 函数 - 函数基础
// 运行方式：go run 01_函数基础.go
// ============================================================

package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println("=== 1. 函数基本用法 ===")
	greet()
	greetUser("张三")
	fmt.Println("add(3, 5) =", add(3, 5))

	fmt.Println("\n=== 2. 多返回值 ===")
	sum, product := calculate(10, 3)
	fmt.Printf("10 和 3: 和 = %d, 积 = %d\n", sum, product)

	// 忽略返回值
	s, _ := calculate(5, 2)
	fmt.Println("只取和:", s)

	fmt.Println("\n=== 3. 命名返回值 ===")
	quotient, remainder := divide(10, 3)
	fmt.Printf("10 / 3: 商 = %d, 余数 = %d\n", quotient, remainder)

	fmt.Println("\n=== 4. 可变参数 ===")
	fmt.Println("sumAll(1,2,3) =", sumAll(1, 2, 3))
	fmt.Println("sumAll(1,2,3,4,5) =", sumAll(1, 2, 3, 4, 5))

	// 传入切片
	nums := []int{1, 2, 3, 4, 5, 6}
	fmt.Println("sumAll(nums...) =", sumAll(nums...))

	fmt.Println("\n=== 5. 递归函数 ===")
	fmt.Println("阶乘:")
	for i := 0; i < 6; i++ {
		fmt.Printf("%d! = %d\n", i, factorial(i))
	}

	fmt.Println("\n=== 6. 函数作为值 ===")
	// 函数变量
	var addFunc func(int, int) int = add
	fmt.Println("函数变量调用:", addFunc(10, 20))

	// 匿名函数
	multiply := func(a, b int) int {
		return a * b
	}
	fmt.Println("匿名函数:", multiply(3, 4))

	fmt.Println("\n=== 7. 闭包 ===")
	counter := makeCounter()
	fmt.Println("闭包计数:", counter()) // 1
	fmt.Println("闭包计数:", counter()) // 2
	fmt.Println("闭包计数:", counter()) // 3

	counter2 := makeCounter()
	fmt.Println("新闭包计数:", counter2()) // 1（独立的计数）

	fmt.Println("\n=== 8. defer 延迟执行 ===")
	fmt.Println("开始")
	defer fmt.Println("延迟执行1")
	defer fmt.Println("延迟执行2")
	fmt.Println("结束")
	// defer 是栈式执行，后声明的先执行

	fmt.Println("\n=== 9. 函数使用示例 ===")
	fmt.Println("斐波那契前 10 项:")
	for i := 0; i < 10; i++ {
		fmt.Printf("%d ", fibonacci(i))
	}
	fmt.Println()

	fmt.Println("\n1 到 50 之间的质数:")
	for num := 1; num <= 50; num++ {
		if isPrime(num) {
			fmt.Printf("%d ", num)
		}
	}
	fmt.Println()
}

// 无参数无返回值
func greet() {
	fmt.Println("Hello, World!")
}

// 带参数
func greetUser(name string) {
	fmt.Printf("你好，%s！\n", name)
}

// 带返回值
func add(a int, b int) int {
	return a + b
}

// 多返回值
func calculate(a, b int) (int, int) {
	return a + b, a * b
}

// 命名返回值
func divide(a, b int) (quotient int, remainder int) {
	quotient = a / b
	remainder = a % b
	return // 自动返回命名的返回值
}

// 可变参数
func sumAll(nums ...int) int {
	total := 0
	for _, n := range nums {
		total += n
	}
	return total
}

// 递归函数
func factorial(n int) int {
	if n == 0 || n == 1 {
		return 1
	}
	return n * factorial(n-1)
}

// 闭包
func makeCounter() func() int {
	count := 0
	return func() int {
		count++
		return count
	}
}

// 斐波那契
func fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	return fibonacci(n-1) + fibonacci(n-2)
}

// 判断质数
func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	for i := 2; i <= int(math.Sqrt(float64(n))); i++ {
		if n%i == 0 {
			return false
		}
	}
	return true
}

File: rust/03_函数/01_函数基础.rs
// ============================================================
// Rust 函数 - 函数基础
// 编译运行：rustc 01_函数基础.rs -o 01_函数基础 && ./01_函数基础
// ============================================================

fn main() {
    println!("=== 1. 函数基本用法 ===");
    greet();
    greet_user("张三");
    println!("add(5, 3) = {}", add(5, 3));
    println!("add(10, 20) = {}", add(10, 20));

    println!("\n=== 2. 函数返回值 ===");
    let five = five();
    println!("five() = {}", five);

    let x = plus_one(5);
    println!("plus_one(5) = {}", x);

    println!("\n=== 3. 表达式 vs 语句 ===");
    // 函数作为表达式
    let y = {
        let x = 3;
        x + 1  // 注意没有分号，这是表达式
    };
    println!("y = {}", y);

    // 语句没有返回值
    // let z = (let x = 3); // 错误！

    println!("\n=== 4. 多参数函数 ===");
    println!("max(5, 10) = {}", max(5, 10));
    println!("max(100, 50) = {}", max(100, 50));

    println!("\n=== 5. 元组返回多个值 ===");
    let (sum, product) = calculate(10, 3);
    println!("10 和 3: 和 = {}, 积 = {}", sum, product);

    // 忽略返回值
    let (s, _) = calculate(5, 2);
    println!("只取和: {}", s);

    println!("\n=== 6. 递归函数 ===");
    println!("阶乘:");
    for i in 0..=5 {
        println!("{}! = {}", i, factorial(i));
    }

    println!("\n=== 7. 闭包 ===");

    // 简单闭包
    let add_one = |x: i32| x + 1;
    println!("add_one(5) = {}", add_one(5));

    let sum = |a, b| a + b;
    println!("sum(3, 4) = {}", sum(3, 4));

    // 捕获环境
    let x = 4;
    let equal_to_x = |z| z == x;
    println!("4 == x: {}", equal_to_x(4));

    println!("\n=== 8. 函数作为参数 ===");
    let numbers = vec![1, 2, 3, 4, 5];
    let doubled: Vec<_> = numbers.iter().map(|&x| x * 2).collect();
    println!("原数组: {:?}", numbers);
    println!("翻倍后: {:?}", doubled);

    println!("\n=== 9. 使用示例 ===");
    println!("斐波那契前 10 项:");
    for i in 0..10 {
        print!("{} ", fibonacci(i));
    }
    println!();

    println!("\n1 到 50 之间的质数:");
    for num in 1..=50 {
        if is_prime(num) {
            print!("{} ", num);
        }
    }
    println!();

    println!("\n=== 运行完成 ===");
}

fn greet() {
    println!("Hello, World!");
}

fn greet_user(name: &str) {
    println!("你好，{}！", name);
}

fn add(a: i32, b: i32) -> i32 {
    a + b  // 表达式，没有分号
}

fn five() -> i32 {
    5
}

fn plus_one(x: i32) -> i32 {
    x + 1
}

fn max(a: i32, b: i32) -> i32 {
    if a > b { a } else { b }
}

fn calculate(a: i32, b: i32) -> (i32, i32) {
    (a + b, a * b)
}

fn factorial(n: u32) -> u32 {
    if n == 0 || n == 1 {
        1
    } else {
        n * factorial(n - 1)
    }
}

fn fibonacci(n: u32) -> u32 {
    if n <= 1 {
        return n;
    }
    fibonacci(n - 1) + fibonacci(n - 2)
}

fn is_prime(n: i32) -> bool {
    if n < 2 {
        return false;
    }
    for i in 2.. {
        if i * i > n {
            break;
        }
        if n % i == 0 {
            return false;
        }
    }
    true
}
